parser/tracing.go

```diff
@@ -26,19 +26,75 @@ func TraceFunction(manager *InstrumentationManager, node dst.Node, tracing *trac
 		panic(fmt.Sprintf("TraceFunction only accepts *dst.FuncDecl or *dst.FuncLit, got %s", nodeType))
 	}
 
-	// add needed tracing object to function declaration parameters
-	// this must be the first thing done, since it double checks the type assignment for tracing and will change it
-	tracingImport, ok := tracing.AddParameterToDeclaration(manager.getDecoratorPackage(), node)
-	if ok {
-		manager.addImport(tracingImport)
-		TopLevelFunctionChanged = true
+	var funcType *dst.FuncType
+	var funcBody *dst.BlockStmt
+
+	// Determine the function type and body based on node type
+	decl, isFuncDecl := node.(*dst.FuncDecl)
+	lit, isFuncLit := node.(*dst.FuncLit)
+
+	if isFuncDecl {
+		funcType = decl.Type
+		funcBody = decl.Body
+	} else if isFuncLit {
+		funcType = lit.Type
+		funcBody = lit.Body
+	} else {
+		// This should not happen due to the initial type check, but it's a safeguard
+		return node, false
 	}
 
-	// create segment if needed
-	segmentImport, ok := tracing.CreateSegment(node)
-	if ok {
-		manager.addImport(segmentImport)
-		TopLevelFunctionChanged = true
+	// Check if the function already has a transaction parameter
+	hasTransactionParam := false
+	for _, param := range funcType.Params.List {
+		for _, ident := range param.Names {
+			if manager.transactionCache.CheckTransactionExists(ident) {
+				hasTransactionParam = true
+				break
```

Comment:
This `break` will only affect the inner loop - the outer loop will continue executing. To avoid that, this logic could be pulled out into a helper function instead where the inner conditional can `return` and the `bool` return value can be used to set `hasTransactionParam`